h5py/h5fd.pyx

```diff
@@ -77,3 +77,106 @@ LOG_TIME_IO   = H5FD_LOG_TIME_IO    # (H5FD_LOG_TIME_OPEN|H5FD_LOG_TIME_READ|H5F
 # Flag for tracking allocation of space in file
 LOG_ALLOC     = H5FD_LOG_ALLOC      # 0x4000
 LOG_ALL       = H5FD_LOG_ALL        # (H5FD_LOG_ALLOC|H5FD_LOG_TIME_IO|H5FD_LOG_NUM_IO|H5FD_LOG_FLAVOR|H5FD_LOG_FILE_IO|H5FD_LOG_LOC_IO)
+
+
+# Implementation of 'fileobj' Virtual File Driver: HDF5 Virtual File
+# Layer wrapper over Python file-like object.
+# https://support.hdfgroup.org/HDF5/doc1.8/TechNotes/VFL.html
+
+# HDF5 events (read, write, flush, ...) are dispatched via
+# H5FD_class_t (struct of callback pointers, H5FD_fileobj_*). This is
+# registered as the handler for 'fileobj' driver via H5FDregister.
+
+# H5FD_fileobj_open callback acts first, expecting "filename" from
+# Python side in form
+#   hex(id(file)).encode('ASCII')
+# and retruning struct H5FD_fileobj_t (descendant of base H5FD_t)
+# which will hold file state. Other callbacks receive H5FD_fileobj_t
+# and operate on f.fileobj. If successful, callbacks must return zero;
+# otherwize non-zero value.
+
+
+# H5FD_t of file-like object
+ctypedef struct H5FD_fileobj_t:
+  H5FD_t base  # must be first
+  PyObject* fileobj
+  haddr_t eoa
+
+
+# A minimal subset of callbacks is implemented. Non-essential
+# parameters (dxpl, type) are ignored.
+
+from cpython cimport Py_INCREF, Py_DECREF
+from libc.stdint cimport *
+from libc.stdio cimport *
+
+
+cdef H5FD_fileobj_t *H5FD_fileobj_open(const char *name, unsigned flags, hid_t fapl, haddr_t maxaddr):
+    cdef unsigned long long int p
+    cdef H5FD_fileobj_t *f
+    if sscanf(name, "%llx", &p):
```

Comment:
But does it mean I'll have to implement the `fapl_*` callbacks, and additionally a way to pass the pointer value to `H5FD_class_t` from Python code? In my personal feeling that will make the code less understandable and so *more* brittle, -- but should be possible technically.

Reply:
right now users can do:
```python
h5py.File('0', driver='fileobj')
```
which will just segfault because it will parse as a pointer and then dereference null. Also, it is not obvious that the object will be alive between the call to `id` and the call to `open`. This happens to be the case now because the driver open call happens in `File.__init__`, and the object must be alive on the stack of the calling frame. Imagine if someone deferred the `open` call to the first use, now nothing has kept the file-like object alive, and `id` may or may not have been reclaimed. This will cause weird bugs later on. By using actual driver arguments we can remove this "serialization" step which makes the required lifetime of the object hard to follow.

Also, on 32 bit systems the pointers will not be long long ints, which means this is a potentially truncating cast.

Reply:
> But what is the workflow implying call ... and what user expects from h5py in this case?

I would expect that code to tell me that `str` is not a file-like object or to get an attribute error or some other Python error. No user ever expects a segfault as the result of writing Python code. A segfault might be okay for very low-level performance sensitive APIs like reading into a custom dataspace where verification of the arguments is expensive relative to the desired performance, but for something like opening a file I would never expect a segfault. Many people who depend on h5py are not software engineers by trade and won't see this as obviously incorrect. These same people will likely not be able to debug the issue 


> it is currently important that open happens shortly after id call

Nothing about this is explicit nor obvious from the code as written. For example, imagine we want to move `open` to the first read for some subclass of `File`, like a `LazyFile` or something. Now, this code will have a bad bug when temporaries are passed to the constructor. I agree that it is easy to ensure that the sequence of events happens correctly: just pass the pointer to the code that needs the pointer.


> A value of type `long long int` is guaranteed to be at least 64-bit wide, so it will be able to contain a pointer on all systems.

I agree that the integer can store the pointer, I am not sure that you can cast the `long long int` back to a 32 bit pointer in a compliant way. The only int types guaranteed to be able to convert to and from pointers are `intptr_t` and `uintptr_t`. The tests are passing but the code may be relying on undefined behavior.

> However please don't expect it to be any more robust or fast in practical cases.

None of this is about speed or performance, it is about maintainability and being explicit with the intent and requirements of code. Passing a pointer from Python to C through a string which is converted to an integer and then cast to a `PyObject*` is non-obvious behavior. There is already an API which allows us to just pass the pointer through to this code without passing it through so many conversions.

Another thing I hadn't thought of until now: in pypy, the `id` might not even be a valid pointer. This would then just cast a random integer to a pointer and go. Idk if pypy is a supported platform, but I see windows pypy wheels for the latest release.

Reply:
The latter argument (pypy support) is pretty valid. I will refactor the code to pass the PyObject as you suggested.